pilot/pkg/model/push_context.go

```diff
@@ -595,8 +614,14 @@ func (ps *PushContext) SubsetToLabels(proxy *Proxy, subsetName string, hostname
 
 // InitContext will initialize the data structures used for code generation.
 // This should be called before starting the push, from the thread creating
-// the push context.
-func (ps *PushContext) InitContext(env *Environment) error {
+// the push context. If updateReq is provided and oldPushContext is not nil,
+// then push context initialization will copy content from oldPushContext for config types
+// that are not in the updateReq, saving some processing time.
+// TODO: This can be expanded to completely avoid pushing to certain namespaces
+func (ps *PushContext) InitContext(env *Environment, oldPushContext *PushContext, updateReq *UpdateRequest) error {
+	// TODO: Why is this lock necessary?
+	// Can two functions simultaneously call init/update on the same push context object? That does not seem to
+	// be the case
```

Comment:
This can happen in 
```
func (s *DiscoveryServer) StreamAggregatedResources(stream ads.AggregatedDiscoveryService_StreamAggregatedResourcesServer) error {
	peerInfo, ok := peer.FromContext(stream.Context())
	peerAddr := "0.0.0.0"
	if ok {
		peerAddr = peerInfo.Addr.String()
	}

	t0 := time.Now()

	// first call - lazy loading, in tests. This should not happen if readiness
	// check works, since it assumes ClearCache is called (and as such PushContext
	// is initialized)
	// InitContext returns immediately if the context was already initialized.
	err := s.globalPushContext().InitContext(s.Env)
```